### PR TITLE
Adding the results of pretest and post test to the monthly custom batch report

### PR DESCRIPTION
**Pull request by felipemontoya**
# Description

This PR basically enhances the CME user dump command to include the grades of all graded sections of the course for the user (During a normal CME course, this sections are of the type Pre-test and Post-test). This feature lives under an option flag `-g` or `--include-grades` which defaults to false.

Notice that the import of the CourseGradeFactory is future proofed to work in your current ficus based release and also in ginkgo based releases as the current master. 

# Reviewers

I tagged you guys here, because we had briefly talked about this briefly in slack. Please let me know if someone else is better for this review.

- [ ] @stvstnfrd 
- [ ] @caesar2164 

# Other Stakeholders

- @juancamilom
- @dfrojas

### CODE REVIEW

**Comment by felipemontoya** [comment]
@stvstnfrd testing this should be simple.
I understand that you already run this command on a monthly basis to generate the data export.
If that is so, then in any environment, dev, stage or even prod (if that is an option), call the CME dump_userinfo command as you normally would.
Then checkout this branch and call the command with the `-g` flag. Both calls should create a similar document, but the one using the flag will add one extra column per graded section and fill it with the grades data.

On devstack, you can use this line to run the command with all the required parameters `python manage.py lms --settings=devstack dump_userinfo --start 2016-12-01 --end 2017-12-01 --course course-v1:edX+DemoX+Demo_Course --outfile /edx/app/edxapp/edx-platform/report.csv --include-grades`

**Comment by juancamilom** [comment]
@stvstnfrd Hi Steve, please be aware of this pending PR, we are confident that it will help the reporting on the CME side and won't affect any of your standard operations.  Plus, we are trying to get this closed on the administrative side before the year is over. 
Thanks for your cooperation. :)

File: lms/djangoapps/instructor/management/commands/dump_userinfo.py
Comment (caesar2164) [inline lms/djangoapps/instructor/management/commands/dump_userinfo.py]: This doesn't seem to be used either?

File: lms/djangoapps/instructor/management/commands/dump_userinfo.py
Comment (caesar2164) [inline lms/djangoapps/instructor/management/commands/dump_userinfo.py]: Also seem unused.

File: lms/djangoapps/instructor/management/commands/dump_userinfo.py
Comment (caesar2164) [inline lms/djangoapps/instructor/management/commands/dump_userinfo.py]: two lines before function declarations...

File: lms/djangoapps/instructor/management/commands/dump_userinfo.py
Comment (caesar2164) [inline lms/djangoapps/instructor/management/commands/dump_userinfo.py]: two lines...

File: lms/djangoapps/instructor/management/commands/dump_userinfo.py
Comment (caesar2164) [inline lms/djangoapps/instructor/management/commands/dump_userinfo.py]: two lines...

**Review by caesar2164 — COMMENTED** [review]
Could you guys run pylint & pep8 on this code?
I'm finding quite a few violations just by looking, but the linters will find all of them...

**Review by caesar2164 — CHANGES_REQUESTED** [review]
We definitely can't merge until it passes linting...

**Comment by caesar2164** [comment]
@juancamilom & @felipemontoya - could one of you also rebase this branch on our current master?

We can't run tests until that happens...

File: lms/djangoapps/instructor/management/commands/dump_userinfo.py
Comment (dfrojas) [inline lms/djangoapps/instructor/management/commands/dump_userinfo.py]: Did not have validation when intervals is smaller than 1 generating a ZeroDivisionError

**Comment by felipemontoya** [comment]
@caesar2164 thanks for your review! We rebased the branch to the latest master, we also checked the local `paver run_quality`. You can run the tests now.

We try to keep the footprint of our changes to a minimum, so we did not organize the imports, nor changed anything that did not prevent us from doing the requested change.

File: lms/djangoapps/instructor/management/commands/dump_userinfo.py
Comment (felipemontoya) [reply lms/djangoapps/instructor/management/commands/dump_userinfo.py]: We went by: method definitions inside a class are surrounded by a single blank line.

Comment (dfrojas) [reply lms/djangoapps/instructor/management/commands/dump_userinfo.py]: Since build_new_columns it's a method of the Command class, has one blank line according to https://www.python.org/dev/peps/pep-0008/#blank-lines

Comment (caesar2164) [reply lms/djangoapps/instructor/management/commands/dump_userinfo.py]: yep you're right, sorry.

File: lms/djangoapps/instructor/management/commands/dump_userinfo.py
Comment (caesar2164) [reply lms/djangoapps/instructor/management/commands/dump_userinfo.py]: if you make the `intervals` definition: 
```
intervals = max(int(0.10 * total), 1)
```
you don't need this if check :)

**Review by caesar2164 — APPROVED** [review]

**Comment by stvstnfrd** [comment]
Their approach is more clear, let's not condense and get clever.

File: lms/djangoapps/instructor/management/commands/dump_userinfo.py
Comment (stvstnfrd) [inline lms/djangoapps/instructor/management/commands/dump_userinfo.py]: I'm a bit concerned/confused with this loop,
specifically that it's nested inside the `for profile in profiles:` loop.

`user_scores` will continue to grow upon each iteration of the profiles list.
So then here, we'll be iterating over more and more users each time,
when it seems like we probably only care about the most recent one, right?

Can we remove this outer `for user in user_scores:` loop and just say:

```python
for exam in user_json['exams']:
    label = exam['label']
    student_dict[label] = exam['score']
```

This would appear to reduce many unnecessary iterations.
Or am I missing something?

**Review by stvstnfrd — CHANGES_REQUESTED** [review]
Most comments are cosmetic.
My only real concern deals with a set of nested for-loops. Take a look and let me know if I'm overlooking anything.

Otherwise, this is looking good.

File: lms/djangoapps/instructor/management/commands/dump_userinfo.py
Comment (dfrojas) [reply lms/djangoapps/instructor/management/commands/dump_userinfo.py]: The conditional is checking the same object that we are building (exam_colums), so here is not possible do a list comprehension

File: lms/djangoapps/instructor/management/commands/dump_userinfo.py
Comment (dfrojas) [reply lms/djangoapps/instructor/management/commands/dump_userinfo.py]: No, is not necessary since `verbosity` is not a mandatory flag, here we are just making sure that `intervals` have a value an avoid a `ZeroDivisionError`

File: lms/djangoapps/instructor/management/commands/dump_userinfo.py
Comment (dfrojas) [reply lms/djangoapps/instructor/management/commands/dump_userinfo.py]: Yes, you are right, it was a naming confusion

File: lms/djangoapps/instructor/management/commands/dump_userinfo.py
Comment (dfrojas) [reply lms/djangoapps/instructor/management/commands/dump_userinfo.py]: `user_scores` was created to store the scores in an independent way, but your solution it's cleaner. Thanks for the observation

File: lms/djangoapps/instructor/management/commands/dump_userinfo.py
Comment (dfrojas) [reply lms/djangoapps/instructor/management/commands/dump_userinfo.py]: Are you meaning the counter? if it is. No, we don't need it.

But we need this loop in order to create the new object in `student_dict` as you propose above.

File: lms/djangoapps/instructor/management/commands/dump_userinfo.py
Comment (dfrojas) [reply lms/djangoapps/instructor/management/commands/dump_userinfo.py]: Sorry, my mistake

File: lms/djangoapps/instructor/management/commands/dump_userinfo.py
Comment (dfrojas) [reply lms/djangoapps/instructor/management/commands/dump_userinfo.py]: Yes, backward-compatibility reason

File: lms/djangoapps/instructor/management/commands/dump_userinfo.py
Comment (dfrojas) [reply lms/djangoapps/instructor/management/commands/dump_userinfo.py]: The tuple is already formed when we use `append`. Already remove the unnecessary parenthesis.

File: lms/djangoapps/instructor/management/commands/dump_userinfo.py
Comment (stvstnfrd) [reply lms/djangoapps/instructor/management/commands/dump_userinfo.py]: Yes, I just meant the counter. Thanks!

File: lms/djangoapps/instructor/management/commands/dump_userinfo.py
Comment (stvstnfrd) [reply lms/djangoapps/instructor/management/commands/dump_userinfo.py]: Ah, gotcha.

File: lms/djangoapps/instructor/management/commands/dump_userinfo.py
Comment (stvstnfrd) [inline lms/djangoapps/instructor/management/commands/dump_userinfo.py]: I'm worried there could be conflicts here due to key collisions.

Since this is a dictionary, If there are two scores with the same label, wouldn't the last in the list overwrite any previous values? e.g. two scores with a label "Quiz".
Or what if one of the labels matches one of the previously existing values, e.g. 'Attested`, etc.?

While possible, this wasn't previously a concern because we (the code) controlled all of the values that ended up in the `student_dict`. But now that the values are pulled dynamically from the sections, it seems like naming collisions are possible and would cause problems.

I think we may need to maintain two different data structures here:
- a dictionary, `student_dict`, could remain as it's needed for the column ordering specified by `CME_SPECIFIC_ORDER`.
- a list, `student_scores`, which is then appended to the CSV row based on the order in which they appear (but allowing for duplicate labels), similar to how it's done below with `build_new_columns`.

Thoughts?

**Review by stvstnfrd — CHANGES_REQUESTED** [review]

File: lms/djangoapps/instructor/management/commands/dump_userinfo.py
Comment (dfrojas) [reply lms/djangoapps/instructor/management/commands/dump_userinfo.py]: The class `AssignmentFormatGrader` found in `/common/lib//xmodule/xmodule/graders.py` initialize a variable named `starting_index` which according to the docstring in that class is "the first number that will appear. For example, starting_index=3 and min_count = 2 would produce the labels "Assignment 3", "Assignment 4"."

If you go to the line 363 you'll see this:

```
short_label = u"{short_label} {index:02d}".format(
                index=i + self.starting_index,
                short_label=self.short_label
            )
```
This concatenates the name of the label and a consecutive. Which means the label will never be repeated.

File: lms/djangoapps/instructor/management/commands/dump_userinfo.py
Comment (dfrojas) [reply lms/djangoapps/instructor/management/commands/dump_userinfo.py]: I'll also move `user_id`

File: lms/djangoapps/instructor/management/commands/dump_userinfo.py
Comment (dfrojas) [reply lms/djangoapps/instructor/management/commands/dump_userinfo.py]: Yes, you are right, I did not notice.

File: lms/djangoapps/instructor/management/commands/dump_userinfo.py
Comment (stvstnfrd) [inline lms/djangoapps/instructor/management/commands/dump_userinfo.py]: Can this label ever actually be empty? Or are we just being overly cautious here?

Do we actually want to omit the value from `csv_fieldnames` or should we include it, but just use an empty/default label instead? Like:

```python
csv_fieldnames = [
    label or '<empty>'
    # ...
]
```

**Review by stvstnfrd — CHANGES_REQUESTED** [review]
One clarifying question, otherwise this needs rebased over the new master. After that, +1

**Comment by dfrojas** [comment]
@stvstnfrd I'm trying to do the rebase, but this file does not exist in master. Any particular reason?

**Comment by dcadams** [comment]
@dfrojas I was moved in a re-factor to
https://github.com/Stanford-Online/edx-platform/blob/master/openedx/stanford/lms/djangoapps/instructor/management/commands/dump_userinfo.py

File: lms/djangoapps/instructor/management/commands/dump_userinfo.py
Comment (dfrojas) [reply lms/djangoapps/instructor/management/commands/dump_userinfo.py]: You mean not check the length of that list and instead try to iterate the list it does not matter if is empty?

**Comment by dfrojas** [comment]
@stvstnfrd The rebase is done. Hope to hear you soon :)

**Review by stvstnfrd — APPROVED** [review]

**Comment by stvstnfrd** [comment]
Hrmm, just seeing some hiccups with the tests, so I'm kicking off a rebuild to take a closer look before merging.

**Comment by stvstnfrd** [comment]
@dfrojas I noticed an issue w/ the rebase, which i went ahead and worked out myself. Please take a look here [1] and let me know if you have any questions. Thanks!

[1] https://github.com/eduNEXT/edx-platform/pull/35

**Comment by stvstnfrd** [comment]
> I did not notice that with the last change, the CME model has changed

Sorry, I don't follow. I'm not seeing a change to model in this PR.
Can you clarify?

> we are ready now for merge?

Yup. I restarted the test build since it failed, but I suspect that's unrelated to your work and should be fine.

**Comment by dfrojas** [comment]
@stvstnfrd Sorry, I meant the array PROFILE_FIELDS. About the test, yes... There was a global error in Circle CI at the time I pushed this change.

**Pull request merged**
